velox/dwio/common/BitUnpacking.h

```diff
@@ -0,0 +1,663 @@
+    _mm256_storeu_si256(reinterpret_cast<__m256i*>(outputBuffer), result);
+
+    inputBuffer += 16;
+    outputBuffer += 8;
+  }
+}
+
+// Unpack numValues number of uint32_t values with bitWidth in [17, 21] range.
+static inline void unpack17to21(
+    uint8_t bitWidth,
+    const uint8_t*& inputBuffer,
+    uint64_t numValues,
+    uint32_t*& outputBuffer) {
+  uint64_t pdepMask = kPdepMask32[bitWidth];
+
+  uint8_t rightShift1 = bitWidth * 2;
+  uint8_t leftShift1 = 64 - rightShift1;
+  uint8_t rightShift2 = bitWidth * 4 - 64;
+  uint8_t rightShift3 = bitWidth * 6 - 64;
+  uint8_t leftShift3 = 128 - bitWidth * 6;
+
+  alignas(16) uint64_t values[4] = {0, 0, 0, 0};
+
+  uint64_t numBytes = (numValues * bitWidth + 7) / 8;
+  auto readEndOffset = inputBuffer + numBytes;
+
+  // Process bitWidth bytes (8 values) a time.
+  while (inputBuffer < readEndOffset) {
+    std::memcpy(values, inputBuffer, bitWidth);
+
+    *reinterpret_cast<uint64_t*>(outputBuffer) = _pdep_u64(values[0], pdepMask);
+    *(reinterpret_cast<uint64_t*>(outputBuffer) + 1) = _pdep_u64(
+        (values[0] >> rightShift1) | (values[1] << leftShift1), pdepMask);
+    *(reinterpret_cast<uint64_t*>(outputBuffer) + 2) =
+        _pdep_u64(values[1] >> rightShift2, pdepMask);
+    *(reinterpret_cast<uint64_t*>(outputBuffer) + 3) = _pdep_u64(
+        (values[1] >> rightShift3) | (values[2] << leftShift3), pdepMask);
+
+    inputBuffer += bitWidth;
+    outputBuffer += 8;
+  }
+}
+
+// Unpack numValues number of uint32_t values with bitWidth in [22, 31] range.
+static inline void unpack22to31(
+    uint8_t bitWidth,
+    const uint8_t*& inputBuffer,
+    uint64_t numValues,
+    uint32_t*& outputBuffer) {
+  uint64_t pdepMask = kPdepMask32[bitWidth];
+
+  uint8_t rightShift1 = bitWidth * 2;
+  uint8_t leftShift1 = 64 - rightShift1;
+  uint8_t rightShift2 = bitWidth * 4 - 64;
+  uint8_t leftShift2 = 128 - bitWidth * 4;
+  uint8_t rightShift3 = bitWidth * 6 - 128;
+  uint8_t leftShift3 = bitWidth * 8 - 128;
+
+  alignas(16) uint64_t values[4] = {0, 0, 0, 0};
+
+  uint64_t numBytes = (numValues * bitWidth + 7) / 8;
+  auto readEndOffset = inputBuffer + numBytes;
+
+  // Process bitWidth bytes (8 values) a time.
+  while (inputBuffer < readEndOffset) {
+    std::memcpy(values, inputBuffer, bitWidth);
+
+    *reinterpret_cast<uint64_t*>(outputBuffer) = _pdep_u64(values[0], pdepMask);
+    *(reinterpret_cast<uint64_t*>(outputBuffer) + 1) = _pdep_u64(
+        (values[0] >> rightShift1) | (values[1] << leftShift1), pdepMask);
+    *(reinterpret_cast<uint64_t*>(outputBuffer) + 2) = _pdep_u64(
+        (values[1] >> rightShift2) | (values[2] << leftShift2), pdepMask);
+    *(reinterpret_cast<uint64_t*>(outputBuffer) + 3) = _pdep_u64(
+        (values[1] >> rightShift3) | (values[2] << leftShift3), pdepMask);
+
+    inputBuffer += bitWidth;
+    outputBuffer += 8;
+  }
+}
+
+// Unpack numValues number of uint16_t values with bitWidth = 16
```

Comment:
use `///`?

Reply:
@leoluan2009 Thanks for reviewing. These functions are not exposed to other compilation units and therefore not public. I'm not sure what the Velox recommendations are. @Yuhta @pedroerp What's  your recommendation on the comment style here?

Reply:
For internal functions we can use `//` for comments.  `///` is for content that you want to be picked up by doxygen.